sf-stream/src/tests/proto_converter_tests.rs

```diff
@@ -0,0 +1,235 @@
+    let mut test_context = new_test_context(current_function_name!(), 0);
+
+    // create user transactions
+    let account = test_context.gen_account();
+    let txn = test_context.create_user_account(&account);
+    test_context.commit_block(&vec![txn.clone()]).await;
+
+    let context = Arc::new(test_context.clone().context);
+    let mut streamer = SfStreamer::new(context, 0, None);
+
+    // emulating real stream, getting first block
+    let converted_0 = streamer.batch_convert_once(1).await;
+    let txn = converted_0.first().unwrap().clone();
+    assert_eq!(txn.version, 0);
+    assert_eq!(txn.type_.unwrap(), TransactionType::GENESIS);
+
+    // getting second block
+    let converted_1 = streamer.batch_convert_once(3).await;
+    // block metadata expected
+    let txn = converted_1[0].clone();
+    assert_eq!(txn.version, 1);
+    assert_eq!(txn.type_.unwrap(), TransactionType::BLOCK_METADATA);
+    if let Txn_data::BlockMetadata(txn) = txn.txn_data.unwrap() {
+        assert_eq!(txn.round, 1);
+    }
+    // user txn expected
+    let txn = converted_1[1].clone();
+    assert_eq!(txn.version, 2);
+    assert_eq!(txn.type_.unwrap(), TransactionType::USER);
+    if let Txn_data::User(txn) = txn.txn_data.unwrap() {
+        assert_eq!(
+            txn.request.payload.type_.unwrap(),
+            PayloadType::SCRIPT_FUNCTION_PAYLOAD
+        );
+        if let Payload::ScriptFunctionPayload(payload) =
+            txn.request.payload.clone().unwrap().payload.unwrap()
+        {
+            let address_str = MoveValue::Address(account.address()).to_string();
+            let address_str = Value::String(address_str).to_string();
+            assert_eq!(*payload.arguments.first().unwrap(), address_str);
+        }
+    }
+    // state checkpoint expected
+    let txn = converted_1[2].clone();
+    assert_eq!(txn.version, 3);
+    assert_eq!(txn.type_.unwrap(), TransactionType::STATE_CHECKPOINT);
+}
+
+#[tokio::test]
+async fn test_block_height_and_ts_work() {
+    let start_ts = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_secs();
+    let mut test_context = new_test_context(current_function_name!(), start_ts);
+
+    // Creating 2 blocks w/ user transactions and 1 empty block
+    let mut root_account = test_context.root_account();
+    let account = test_context.gen_account();
+    let txn = test_context.create_user_account_by(&mut root_account, &account);
+    test_context.commit_block(&vec![txn.clone()]).await;
+    let account = test_context.gen_account();
+    let txn = test_context.create_user_account_by(&mut root_account, &account);
+    test_context.commit_block(&vec![txn.clone()]).await;
+    test_context.commit_block(&[]).await;
+
+    // key is version and value is block_height
+    let block_mapping = HashMap::from([
+        (0, 0),
+        (1, 1),
+        (2, 1),
+        (3, 1),
+        (4, 2),
+        (5, 2),
+        (6, 2),
+        (7, 3),
+        (8, 3),
+    ]);
```

Comment:
This is very very very fragile.  It's dependent on whether the number of transactions in a block don't change.

Additionally, this doesn't match reality, the first block after genesis is 1 txn and is a reconfig.

Reply:
I'm confused sorry why is this not guaranteed? I'm mocking the transactions then confirming that the block height indeed matches. This will only break if we decide that a transaction isn't 
block_metadata
user, user, user, etc.
state_checkpoint